### PR TITLE
Add Comments Feature 

### PR DESCRIPTION
**Pull request by roberthuskins**
Screen Shot 2020-02-29 at 11 47 52 PM
SPS Week 2 (Steps 2-5)

Portfolio Site now allows user to post comments using GCP Datastore.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -19,14 +19,60 @@
...
   }
+
+    private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+        String value = request.getParameter(name);
+        if (value == null) {
+            return defaultValue;
+        }
+        return value;
+    }
+    private String convertToJsonUsingGson(ArrayList<String> a) {
```
Comment (IngridFeng) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: convertToJson should be enough for a function name, you can put doc on the line above this function explaining how you did the conversion.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -19,14 +19,60 @@
...
+    datastore.put(newComment);
+    response.sendRedirect("/index.html");
   }
+
+    private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+        String value = request.getParameter(name);
+        if (value == null) {
+            return defaultValue;
+        }
+        return value;
```
Comment (FluffySheep) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Can be simplified to 
```
return value == null ? defaultValue : value;
```

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -12,17 +12,30 @@
...
-  greetingContainer.innerText = greeting;
+  const project = projects[Math.floor(Math.random() * projects.length)];
+  window.open(project, '_blank');
 }
+
+function getComments() {
+    output = "";
+    fetch('/data')  // sends a request to /my-data-url
+    .then(response => response.json()) // parses the response as JSON
+    .then((myObject) => { // now we can reference the fields in myObject!
```
Comment (IngridFeng) [inline portfolio/src/main/webapp/script.js]: nit: indentation

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -12,17 +12,30 @@
...
+    var i;
+ 
+    for (i =0; i < myObject.length; i++) {
+        output += myObject[i];
+        output += "\n";
```
Comment (IngridFeng) [inline portfolio/src/main/webapp/script.js]: nit: using i here should be okay but you might want more descriptive variable name than myObject

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,65 +14,53 @@
...
-        for (Entity entity : results.asIterable()) {
-            comments.add((String)entity.getProperty("text"));
-        }
-        System.out.println(comments);
-        System.out.println("Just printed out comments arr");
-        String json = convertToJsonUsingGson(comments);
-        response.setContentType("application/json;");
-        response.getWriter().println(json);
+    ArrayList<String> comments = new ArrayList();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (IngridFeng) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Initialize this directly inside class. Declare this in init(). 
So like directly under `public class DataServlet extends HttpServlet`.
```
DatastoreService datastore;
@Override
public void init() {
  datastore = DatastoreServiceFactory.getDatastoreService();
}
```
And then you can also delete this line in doPost.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -12,17 +12,30 @@
...
+  const project = projects[Math.floor(Math.random() * projects.length)];
+  window.open(project, '_blank');
 }
+
+function getComments() {
+  output = '';
+  fetch('/data')                          
+      .then(response => response.json()) 
+      .then((jsonData) => {  
+        var i;
```
Comment (FluffySheep) [inline portfolio/src/main/webapp/script.js]: This should be defined in your for loop

```
for (var i = 0; i < jsonData.length; i++)
```

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -12,17 +12,30 @@
...
 
-  // Add it to the page.
-  const greetingContainer = document.getElementById('greeting-container');
-  greetingContainer.innerText = greeting;
+  const project = projects[Math.floor(Math.random() * projects.length)];
+  window.open(project, '_blank');
 }
+
+function getComments() {
+  output = '';
```
Comment (FluffySheep) [inline portfolio/src/main/webapp/script.js]: output should be defined right outside of the forloop it is used in. Keep variables close to where they are used unless it causes overhead.

**Pull request merged**